### DISCUSSION TITLE
Dead Data Sources - What to do when there is no more data?

### DISCUSSION THREAD

**Discussion by alepping** [nebulastream/nebulastream#169] [Ideas]
# Description
(The following is mainly based on discussions with @ankitgit and @ls-1801.)
While streaming sources may run indefinitely, in reality, they might disconnect, they might want to go out of service for an indefinite amount of time, or they might want to say that they have no more data, e.g., end of file of a file source.
Currently, if a single file source detects that it read a complete file, it triggers a soft end of stream, terminating the entire query that the file source belongs to (or potentially all associated queries, if it is a shared source?).

# The Problem
**P1**: A potential problem arises if a user wants to, e.g., union 100 sources and the user does not care about how many of these sources produce data, but the user simply wants to process all available data. Currently, a single source that decides that it has finished reading data terminates the entire query.

# Stream Only (no batch sources)
Since combining batch and streaming sources adds additional complexity, we only consider streaming sources in this discussion.

## Main Question
What is the expected behavior if a source stops producing data?
Potential reasons:
External:
- user/coordinator stops a query 
  - stops an entire decomposed query (not addressed in the result of this discussion)
    - would change if there was a 'stopSource(originId)' request
Internal:
- bad source configuration
- disconnect
- end of stream

## Assumption
We assume that the changes described in #69, and in particular in #71 have already been applied. As a result, the logic to start and stop queries is moved out of the data sources. If a data sources cannot process any more data, it propagates this information upwards, to the query manager, using error codes as described in #32.

## Current Behavior - A0
If a source decides that it cannot read any more data, it triggers a soft end of stream, leading to a reconfiguration message that stops the entire query (or potentially multiple queries).
### Advantages
1. simple to reason about, because behavior is clearly defined
2. simple to log, simply log 'source X stopped because of reason Y'
3. transparent to user, reason for query termination is part of the log and clearly understandable
4. already implemented
### Disadvantages
1. A single source *S1* that stops producing data terminates a query that might not be dependent on S1 and that potentially processes data of hundreds of sources (P1)
2. No flexibility, users or the system cannot decide if a query depends on a source and should therefore stop if the source stops

## Alternative Behavior - A1
A query only stops if the entire query triggers a stopping policy, e.g., no more data since amount of time T, state grows to large, critical number of sources are 'dead'. That is, only the results of the entire query can trigger an automatic stop, not the results of a single source. Requires for state of sources to be propagated upwards (QueryManager, Coordinator, User).

### Advantages
1. High flexibility, users can change a default behavior and customize it to their needs
2. solves P1
### Disadvantages
1. Adds complexity to query processing (what about sources that produce data very infrequently)
2. Flexibility might lead to confusion
3. Bad configuration could lead to dead queries that are still running

## Alternative Behavior - A2
Queries are only stopped by the user.
### Advantages
1. simple logic, easy to implement
2. solves P1

### Disadvantages
1. currently, 'dead' sources would still consume a thread that could be used for useful work
2. potentially less transparent to the user that a query is 'dead' (no more data vs specific query termination log message)

## Alternative Behavior - A3
After a data source decides that it cannot process data anymore, the QueryManager triggers a process that notifies downstream operators. The downstream operators can then decide whether the missing data is critical or not. If it is critical, the operator triggers an end of stream.
```ocaml
(* Example: *)
TCPSource(disconnected) --> Filter --> JsonFormatter --> FileSink
(* the TCPSource disconnected, the Filter is entirely dependent on the TCPSource and would therefore trigger a query stop *)

CSVSource(exhausted) --(build)-----\
                                     --------> Join -> JsonFormatter -> FileSink
TCPSource(running) ------(probe)----/
(* here, after building, the Join can still produce data as long as the TCPSource is running and might therefore not stop the query *)
```

### Advantages
1. could automatically add support for batch sources (CSV, external databases, etc.), if we built state using the data sources
  - potentially users could configure an update frequency used to update the state created from the batch source
2. solves P1 (a union could decide to not terminate the query if other sources still produce data
3. avoids complex system configuration on the user side (in contrast to A1)

### Disadvantages
1. requires more complex operator logic
2. query behavior becomes more complex and potentially less transparent to user
3. less flexible than A1, but could be combined, e.g., a user could overwrite the behavior described in A3 with configurations (e.g., stop query if at least one source dies)
```

**Comment by PhilippGrulich** [comment]
Hi,
I think this is a very important discussion, but I have the feeling that the points and solutions you sketch here are a bit too unspecific. 
I think what you mainly describe here is tailored to the batch processing case, where sources can terminate. 
In a stream processing case, this can only happen in the case of an error. 
This is also the case for the current solution, which uses end-of-stream messages. This design was historically driven by the need to test the system using a batch model while ensuring a happen-before relationship between the data and reconfiguration tasks. 
Furthermore, I don't completely agree with your assessment of the disadvantages:

1: A single source S1 that stops producing data terminates a query that might not be dependent on S1 and that potentially processes data of hundreds of sources (P1) -> 
In the case of a soft-end-of-stream message, each pipeline will receive the message and reduce its active producer counter. Only if the counter reaches zero, it will emit the soft-end-of-stream to the next downstream pipeline. Thus, it is possible that only one source terminates, while other sources still produce data. In the case of a hard-end-of-stream (e.g., caused by an error) the stop is directly propagated, and the query stops. 

2. No flexibility, users or the system cannot decide if a query depends on a source and should therefore stop if the source stops -> 
This can be controlled via the end-of-stream message type, which is inserted at the source. You could also introduce more end-of-stream types. I think, we defined some other cases in the design from Johannes. 

One important aspect is that you also have to ensure correctness while stopping a query. 

------------------
I think, you actually really want to have a proper model for query processing that discusses both streaming (main use-case) and batch queries (at least for testing). I know this sounds like a large topic, but I think it is very hard to avoid, and will also make many things much more clean. 

Maybe a good first step would be to define a very clean query model that defines the entities sources, queries, sinks, steams, and tables, their properties, e.g., continuous sources vs. finite sources, and the connection between these independent entities.

Based on this model, you can then derive the actual requirements for the physical entities and answer questions like:
- What happens with queries if a continuous source terminates, because of an error?
- What happens if a finite source terminates, e.g., as part of a test.
- What are the correctness guarantees you provide? 
- How can a user terminate a query? 
- ...

Some half-baked thoughts:
- A continuous source (KAFKA, MQTT) can only be terminated if the query terminates, or it has an error.
- A finite source (FILE, TABLE_SCAN) terminates and emits soft-end-of-stream messages.

**Reply by alepping** [reply]
Hey @PhilippGrulich, 
I completely agree that the solution sketches are very rough. The idea of this discussion was to work towards a well-defined solution.
> In a stream processing case, this can only happen in the case of an error.

True, but there are many different reasons for 'errors' and errors should potentially be considered the norm.
> Only if the counter reaches zero, it will emit the soft-end-of-stream to the next downstream pipeline.

This is not what I have been observing, but this might be a problem of the CSV source and it might make sense in case of the CSV source, because it is a batch source. If the CSV source detects an EOF, it triggers a soft-end-of-stream that terminates the query, potentially before a TCP source of the same query even connected.
I need to do further investigation concerning streaming sources.
> In the case of a hard-end-of-stream (e.g., caused by an error) the stop is directly propagated, and the query stops.

Is this true, if a single source runs into an error? So an error detected by a data source that prevents the source from reading further data triggers a hard end of stream?
> One important aspect is that you also have to ensure correctness while stopping a query.

I completely agree, also with the next assessment that a proper model for query processing is required and I hope this discussion can lead there.
>  a very clean query model that defines the entities sources, queries, sinks, steams, and tables, their properties, e.g., continuous sources vs. finite sources, and the connection between these independent entities.

Thanks for the list, this is a good starting point! The follow-up questions and half-baked thoughts also help. I will try to concretize the problem next.

**Comment by adrianmichalke** [comment]
I agree here with Philipp, that the problem might resolve when we cleanly differentiate between batch and stream sources.

More half-baked thoughts:

End-of-stream messages should only be issued by finite sources.
- Batch query (only finite sources): ends after got all end-of-streams by all sources
- Stream query (only infinite sources): no end-of-stream
- Batch/Stream (at least one finite and one infinite source): ignores end-of-stream of finite sources (cannot find a use case where something else is desired, help :)

I would handle errors different to end-of-stream messages as the design space gets now way more complex. The expected behavior might be different from source to source and query to query and error to error. I shamelessly plug here my discussion sketch on error streams #153 , where the idea is to let the user handle the errors by writing queries on a stream of the errors.

**Comment by PhilippGrulich** [comment]
Yes makes sense.
I think originally we had.
Soft-end-of-stream:
This basically means that the source terminates but we will process all the data up to this point and produce a consistent and deterministic result.
Hard-end-of-steam: 
Stop as fast as possible and I don't care about the results.

I'm not sure if this behavior is correctly implemented everywhere.
